New Filtering System

mbaruh opened the pull request:
This pull request applies the necessary changes to the database and API to support the new filtering system on the bot.

For more information on how the information in these tables will be used, check out the bot PR: https://github.com/python-discord/bot/pull/2390

> **Warning** This PR must be merged at the same time as the bot PR.

closes #479.

## Scheme
![image](https://user-images.githubusercontent.com/18114431/215279243-58736f06-e76a-421c-82cb-b01f1302f45d.png)

* There is a unique constraint on combinations of name, list_type in FilterList.
* There is a unique constraint on combinations of content, additional_field, filter_list and the settings (everything below filter_list) in Filter.
  * This uses the NULLS NOT DISTINCT feature introduced in psql 15, as two filters in the same filter list with null for some setting have in practice the same value (the value for that setting will be taken from the same filter list).

This scheme is different from the one in the related issue, and is the result of implementing the suggestions made in #482.

## Serializers
The two models use custom serializers in order to provide the settings in a format the bot can work with.
* All settings are placed under one field, `"settings"`.
* Some settings are grouped together, e.g the four settings relating to channels and categories are under the `"channel_scope"` group.

Akarys42 committed:
Filter: new schema

This commit adds new filter schema as described in #479

Akarys42 committed:
Filters: Add new models to Django Admin

Akarys42 committed:
Filters: Add migration to the new model

This will take the currently defined filter list and put them inside the new schema while trying to keep defaults similar to our current setup.

Akarys42 committed:
Filters: hook the new models into the REST API

Akarys42 committed:
Filters: update tests to the new schema

Akarys42 committed:
Filters: allowed_category -> allowed_categories

kosayoda committed:
Improve help text message.

kosayoda committed:
Fix faulty model enumeration.

This also allows us to simplify the str dunder for a FilterList.

kosayoda committed:
Improve name of dm sent to triggered user.

kosayoda committed:
Migrate misc field names and help text changes.

kosayoda committed:
Correct Filter-FilterList relationship.

Instead of a many-many relationship, one filterlist has multiple
filters. Nested serialization is read-only by default, so not
all CRUD methods are implemented yet for the FilterList viewset.

ks129 committed:
Remove one-to-one relationships from filters tables

ks129 committed:
Remove old one-to-one filters relationships serializers, views and URLs

ks129 committed:
Add AbstractModelMeta mixin

ks129 committed:
Update filters migrations

ks129 committed:
Remove default field from FilterSettingsMixin and migration

ks129 committed:
Remove old models from migration

ks129 committed:
Update filters API to actually work

ks129 committed:
Add basic validation for infraction fields + use common infraction types

ks129 committed:
Add validation to filters to not allow duplicated channels and categories

ks129 committed:
Add validation to filters to not allow duplicates + additional_field -> JSON

ks129 committed:
Move filters validations to serializers

ks129 committed:
Add merge migration

ks129 committed:
Fix linting

D0rs4n committed:
Adjust Filter JSON Schema

From now on the Serializer will have a different JSON representation than the
table schema itself, conforming to the format needed on the bot-side.

D0rs4n committed:
Adjust FilterList Representation

From now on the FilterList Serializer will contain a settings field with all the settings that were listed previously, on the model.

D0rs4n committed:
Adjust FilterList Schema to group settings into subcategories

- This commit patches the FilterList serializer's schema, and puts the settings into the relevant subcategories.

D0rs4n committed:
Patch Filter and FilterList Serializer validation logic and representation

- This commit patches an error with the FilterListSerializer validation logic, so that it won't raise an error when an optional field is not present.
- It also adds the `enabled` and `delete_messages` fields, to the FilterSerializer's representation
- Furthermore the commit introduces minor bug patches, regarding DRF Serializer Fields.

D0rs4n committed:
Prepare FilterList and Filter models, serializers for the new filter schema

- Rename channel scope fields:
 - "allowed" -> "disabled"
eg.: "allowed_channels" -> "disabled_channels"
- Rename FilterLists` names:
filter_token -> tokens
domain_name -> domains
guild_invite -> invites
file_format -> formats
- Patch the docs and validators accordingly.

D0rs4n committed:
Patch Filter/FilterList's default values and add new fields

- Patch default values, so that further implementations can be performed on the bot side
- Add three new fields: "send_alert", and in settings under the "server_message" field: "send_message_text", and "server_message_embed" fields.
- Patch documentation, and validators accordingly.
- Perform further patches, and minor corrections.

D0rs4n committed:
Include 'dm_content ' field under Infraction settings in Filters/FilterLists

D0rs4n committed:
Correct 'Redirect' FilterLists' default values.

mbaruh committed:
Merged infraction and notification settings in JSON

The settings for infracting and notifying the user were merged under one field, which is renamed to "infraction_and_notification".
The only place which sends a message in the server by default is the antimalware, the rest try to DM the user first, and antimalware can do the same.

This avoids complications which may result from the filtering cog trying to send two messages: one for the defined server message, and another for a failed DM.

mbaruh committed:
Merge branch 'main' into new-filter-schema

mbaruh committed:
Merge migrations and correct filter defaults to be all null

mbaruh committed:
Allow filter descriptions to be null

mbaruh committed:
Use singular nouns for filter list names

D0rs4n committed:
Patch a minor issue with FilterList field naming in migrations

mbaruh committed:
Changed channeld fields to char arrays, merged migrations

- The fields concerning channels were changed to contains strings instead of integers in order to allow specifying channels and categories by name.

The migrations were merged into a single migration.

mbaruh committed:
Extensions list is ALLOW, not DENY

mbaruh committed:
Create placeholder value for dm embed content in ext list

Some value is needed to signal the bot a message should be sent for a blocked extension. The value itself will be changed at runtime, but this allows avoiding the bot code delving into the exact API response format.

mbaruh committed:
Add invites denylist to the migration

mbaruh committed:
Refine DM content for domains

mbaruh committed:
Change ping fields names

mbaruh committed:
Remove role validation

Roles can be either IDs or names, so the current validation is not relevant anymore. Furthermore the ping fields can accept user IDs or names.

D0rs4n committed:
Sync Filter models with relating migrations, adjust code consistency

D0rs4n committed:
Add UniqueConstraint to the Filter model

- The UniqueConstraint includes every field, except for id and description.

mbaruh committed:
Merge branch 'main' into new-filter-schema

mbaruh committed:
Add voice mute to infraction choices

mbaruh committed:
Use the new models.JSONField

mbaruh committed:
Convert the infraction choices to uppercase

This is done to match the values used on the bot.

mbaruh committed:
Fix send_alert not being added correctly in serializers

mbaruh committed:
Bring back enabled categories

There needs to be a way to only enable a filter in a specific category, so this setting now fulfills that role. Disabled channels can be used to disable a filter in a specific channel within the category.

mbaruh committed:
Merge branch 'main' into new-filter-schema

mbaruh committed:
Properly add dm_embed to serializers

mbaruh committed:
Add infraction channel setting

mbaruh committed:
Fix categories validation

mbaruh committed:
Allow ping arrays to be empty

mbaruh committed:
Allow char fields to be blank

This is necessary allow filters to define a blank message when the default is not blank.

Additionally allows bypass_roles to be empty like the other array fields

mbaruh committed:
Refactors filters serialier

mbaruh committed:
Add uniques filter list to migrations

mbaruh committed:
Add antispam list in migrations

mbaruh committed:
Stop using None as a valid setting value

See e100ae9b on bot

mbaruh committed:
Add creation and update timestamps to filtering models

This is to support auto-infractions reporting (bot 7fcec400)

mbaruh committed:
Disable everyone filter in code jam categories

This is in line with what already existed.

mbaruh committed:
Rename delete_messages to the more generic remove_context

mbaruh committed:
Fix filter serializers validation to account for filterlist settings

mbaruh committed:
Update tests

mbaruh committed:
Merge branch 'main' into new-filter-schema

mbaruh committed:
Make filter unique constraint use NULLS NOT DISTINCT

The existing constraint was ineffective as null values were considered distinct, and so two filters with the same content and no overrides were considered different.

This change uses a new PSQL 15 feature unsupported in django currently, and so it is added with raw SQL.

mbaruh committed:
Update viewsets, fix linting

wookie184 commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,260 @@
+    infraction_duration = models.DurationField(
+        null=False,
+        help_text="The duration of the infraction. Null if permanent."
+    )
```
"null if permanent" is impossible given "null=False`

I see the difficulty since for `FilterBase` null means "not overwritten" so we can't really make use of it for other meanings. I guess you've already sorted this on the bot side so the help text just needs changing

mbaruh replied:
Good spot, should be 0. 4eecda9

wookie184 commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,260 @@
+        null=True
+    )
+
+    # Check FilterList model for information about these properties.
```
I'm not sure I see anything in the `FilterList` model that isn't here? Help text here seems fine though, fine to leave exact specification to the bot side.

mbaruh replied:
This code has gone through so many iteration I have no clue why this was written. 4eecda9

wookie184 commented on pydis_site/apps/api/viewsets/bot/filters.py:
```diff
@@ -0,0 +1,499 @@
+
+    #### Status codes
+    - 204: returned on success
+    - 404: if a tag with the given `id` does not exist
```
```suggestion
    - 404: if a Filter with the given `id` does not exist
```

mbaruh replied:
4eecda9

wookie184 commented on pydis_site/apps/api/migrations/0086_new_filter_schema.py:
```diff
@@ -0,0 +1,171 @@
+import django.db.models.deletion
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+
+import pydis_site.apps.api.models.bot.filters
```
Is there a reason this isn't just `import pydis_site.apps.api.models` since that's what's being accessed? Same for other migrations.

Or just a `from` import to get the names directly?

mbaruh replied:
I have no idea, maybe because it was a semi-generated migration 4eecda9

wookie184 commented on pydis_site/apps/api/viewsets/bot/filters.py:
```diff
@@ -0,0 +1,499 @@
+
+    #### Status codes
+    - 204: returned on success
+    - 404: if a tag with the given `id` does not exist
```
```suggestion
    - 404: if a FilterList with the given `id` does not exist
```
Caught red handed!

mbaruh replied:
👀 4eecda9

wookie184 reviewed (changes requested):
I've not looked at the bot side yet so this is just comments while I was understanding how things work. Looks really good though.

GDWR commented on pydis_site/apps/api/serializers.py:
```diff
@@ -141,30 +144,270 @@ class Meta:
+    """Get the value directly from the key, or from the filter list if it's missing or is None."""
+    if data.get(field_name):
+        return data[field_name]
+    return getattr(data["filter_list"], field_name)
```
```suggestion
def get_field_value(data: dict, field_name: str) -> Any:
    """Get the value directly from the key, or from the filter list if it's missing or is None."""
    if data.get(field_name) is not None:
        return data[field_name]
    return getattr(data["filter_list"], field_name)
```

This should be checked to be `None`, to avoid a value being falsy or even `False`. Take the example of a filter with the value `enabled=False`. in this case, the `filter_list`'s `enabled` field would be used instead.

ChrisLovering replied:
wouldn't we want `is not None` here? So if it's not None return it, else return the filter list default

GDWR replied:
Maybe, lol. You are correct. I've amended the suggestion. Thankyou.

mbaruh replied:
Good spot [be854fa](https://github.com/python-discord/site/pull/861/commits/be854fa3d34dac7b4b9e96b3736dd61d972f1b79)

mbaruh committed:
Fix documentation and import

Co-authored-by: wookie184 <wookie1840@gmail.com>

mbaruh committed:
Fix filter serializers for false-y values

Co-authored-by: GDWR <gregory.dwr@gmail.com>

shtlrs commented on pydis_site/apps/api/migrations/0087_unique_constraint_filters.py:
```diff
@@ -0,0 +1,25 @@
+    ]
+
+    operations = [
+        migrations.RunSQL(
```
Shouldn't we make this a reversible operation ?

I say that because all the default migrations are mostly reversible, and for this case, we'd need to provide a an sql script to the `reverse_sql` kwarg

mbaruh replied:
How would I reverse it?

shtlrs replied:
It should be done by adding `reverse_sql="ALTER TABLE api_filter DROP CONSTRAINT unique_filters"` kwarg to the `RunSQL` instance.

The reason I'm asking this is because when testing locally, and then wanting to move to a different branch to test something else, you'll need to apply migrations up until some version prior to yours, which will result in a 
`django.db.migrations.exceptions.IrreversibleError`

This is ok to happen locally, as you can always drop the container & spin up a new one. 

It's also fair to day that I'm not sure this is something that'll ever happen in production, so we might see what others have to say about it before we action it.

mbaruh replied:
I added it anyway, it can't hurt [1385060](https://github.com/python-discord/site/pull/861/commits/138506045acd1cf464dc63d4465d36208801a3ab)

However if you need to switch to another branch then I'm not sure if some of the other migrations are reversible. I use a separate clone for this project.

shtlrs replied:
I was referring to all the filter migrations, but I don't think this'll ever be a concern.
Thanks for doing this already, and i think it's fine to skip the rest.

shtlrs commented on pydis_site/apps/api/migrations/0088_unique_filter_list.py:
```diff
@@ -0,0 +1,102 @@
+        filter_dm=False,
+        remove_context=True,
+        dm_content=(
+            "I noticed you posted a seemingly valid Discord API "
```
Looks like the bot is self aware :p

shtlrs commented on pydis_site/apps/api/migrations/0089_antispam_filter_list.py:
```diff
@@ -0,0 +1,52 @@
+
+
+def create_antispam_list(apps: Apps, _):
+    """Create the 'unique' FilterList and its related Filters."""
```
```suggestion
    """Create the 'antispam' FilterList and its related Filters."""
```

mbaruh replied:
[d656584](https://github.com/python-discord/site/pull/861/commits/d6565845ec07b9d76949da36505b3f53e402d230)

shtlrs commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,259 @@
+            ),
+        ),
+        help_text="Channel in which to send the infraction.",
+        null=False
```
Just curious, are you passing `False` to  the `null` kwarg for explicitness ? because they default to `False` IIRC

mbaruh replied:
I honestly don't remember.

shtlrs commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,259 @@
+    )
+    disabled_channels = ArrayField(
+        models.CharField(max_length=100),
+        help_text="Channels in which to not run the filter."
```
```suggestion
        help_text="Channels in which to not run the filter even if it's enabled in the category"
```
I suppose ?

mbaruh replied:
[d656584](https://github.com/python-discord/site/pull/861/commits/d6565845ec07b9d76949da36505b3f53e402d230)

shtlrs commented on pydis_site/apps/api/serializers.py:
```diff
@@ -141,30 +144,270 @@ class Meta:
+MENTIONS_FIELDS = ("guild_pings", "dm_pings")
+
+
+def _create_filter_meta_extra_kwargs() -> dict[str, dict[str, bool]]:
```
We can probably condense this one & `_create_filter_list_meta_extra_kwargs` like this
```py
def _create_meta_extra_kwargs(for_filter_list: bool) -> dict[str, dict[str, bool]]:
    """Create the extra kwargs of the for either FilterList serializer's Meta class."""
    extra_kwargs = {}
    for field in SETTINGS_FIELDS:
        field_args = {} if for_filter_list else field_args = {'required': False, 'allow_null': True}
        if field in ALLOW_BLANK_SETTINGS:
            field_args['allow_blank'] = True
        if field in ALLOW_EMPTY_SETTINGS:
            field_args['allow_empty'] = True
        extra_kwargs[field] = field_args
    return extra_kwargs
 ```

and then call them with the correct value of `for_filter_list` in the meta classes

mbaruh replied:
[Merge the extra kwargs creation functions](https://github.com/python-discord/site/pull/861/commits/43913623f87329b51cd2a6793e843c30368698aa)

shtlrs commented on pydis_site/apps/api/serializers.py:
```diff
@@ -141,30 +144,270 @@ class Meta:
+            & set(get_field_value(data, "enabled_channels"))
+        ):
+            raise ValidationError(
+                "You can't have the same value in both enabled and disabled channels lists."
```
It would be painful to send two big lists of enabled & disabled channels that **might** have the common values.

Wouldn't it be better to return the ids that were found in both of them ?

So something like

```py
        if (
            common_channels := set(get_field_value(data, "disabled_channels"))
            & set(get_field_value(data, "enabled_channels"))
        ):
            raise ValidationError(
                f"You can't have the same value in both enabled and disabled channels lists: {list(common_channels)}"
            )
```

Same would go for categories

mbaruh replied:
[Specify the common elements in the validation errors](https://github.com/python-discord/site/pull/861/commits/36bca58ff336f9d4b797a2c76f08775f9de7e9a7)

shtlrs reviewed (changes requested):
This looks superb ! 
Just a couple of questions & nitpicks here & there.

I have done some basic api testing and I haven't really noticed anything just yet.
I'll come back to approve this once I'm done with the bot testing part :D

jchristgit commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,259 @@
+        null=False
+    )
+    bypass_roles = ArrayField(
+        models.CharField(max_length=100),
```
Shouldn't this be a big integer field - because we're dealing with snowflakes?

mbaruh replied:
Roles can be specified as names as well.

jchristgit commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,259 @@
+        null=False
+    )
+    guild_pings = ArrayField(
+        models.CharField(max_length=100),
```
Shouldn't this be a big integer field - because we're dealing with snowflakes?

mbaruh replied:
Pings can be specified as role names as well.

jchristgit replied:
Is a ping a user mention? Those can definitely change. Or what are we referencing in here?

mbaruh replied:
It's either an ID or a role name. The meaning is resolved by the bot.

jchristgit commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,259 @@
+    )
+    # Where a filter should apply.
+    enabled_channels = ArrayField(
+        models.CharField(max_length=100),
```
Same snowflake comment applies here and to the following three fields.

mbaruh replied:
Channels and categories can be specified by their names as well.

jchristgit replied:
Names can change. What about a filter on one of the off-topic channels?

mbaruh replied:
Then for those you can use the channel ID. Name support was added to match the current filters which don't trigger in categories called "CODE JAM". I expect that we will mostly use IDs, but it might come in handy in additional scenarios in the future.

jchristgit commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,259 @@
+        help_text="Channel in which to send the infraction.",
+        null=True
+    )
+    guild_pings = ArrayField(
```
Here the same questions as to BigIntegerField applies as with the model above.

mbaruh replied:
Same as the other guild_pings

jchristgit commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,259 @@
+            UniqueConstraint(
+                fields=tuple(
+                    [field.name for field in FilterBase._meta.fields
+                     if field.name not in ("id", "description", "created_at", "updated_at")]
```
Nitpick, you don't need the `[]` here and can avoid building the intermediate list :)
```suggestion
                fields=tuple(
                    field.name for field in FilterBase._meta.fields
                     if field.name not in ("id", "description", "created_at", "updated_at")
```

mbaruh replied:
[Use consistent quoting style](https://github.com/python-discord/site/pull/861/commits/91b89475913210400cf39884efe37ab5552efbf7)

jchristgit commented on pydis_site/apps/api/serializers.py:
```diff
@@ -141,30 +144,270 @@ class Meta:
+BASE_SETTINGS_FIELDS = (
+    "bypass_roles",
+    "filter_dm",
+    "enabled",
```
There is some inconsistency here with the quotation style between above and below. Personally I would think the fields here and below should be single quoted as well, or is there a reason why these differ?

mbaruh replied:
I think it was just created by different people. I don't mind aligning all of them to one style or another.

mbaruh replied:
What makes you prefer single quotes over double quotes here?

jchristgit replied:
I use single quotes for "data" like dictionary keys, and double quotes for longer user-facing strings like log messages. But it's really personal preference: doesn't matter much

mbaruh replied:
[Use consistent quoting style](https://github.com/python-discord/site/pull/861/commits/91b89475913210400cf39884efe37ab5552efbf7)

jchristgit commented on pydis_site/apps/api/serializers.py:
```diff
@@ -141,30 +144,270 @@ class Meta:
+    """Get the value directly from the key, or from the filter list if it's missing or is None."""
+    if data.get(field_name) is not None:
+        return data[field_name]
+    return getattr(data["filter_list"], field_name)
```
A tiny bit more quoting inconsistency here.

mbaruh replied:
[Use consistent quoting style](https://github.com/python-discord/site/pull/861/commits/91b89475913210400cf39884efe37ab5552efbf7)

jchristgit commented on pydis_site/apps/api/serializers.py:
```diff
@@ -141,30 +144,270 @@ class Meta:
+
+        if (
+            set(get_field_value(data, "disabled_categories"))
+            & set(get_field_value(data, "enabled_categories"))
```
I believe these should be single quotes.

mbaruh replied:
[Use consistent quoting style](https://github.com/python-discord/site/pull/861/commits/91b89475913210400cf39884efe37ab5552efbf7)

jchristgit reviewed (changes requested):
This looks super good. Thank you for your hard work.

I have some smaller comments: they are mostly related to code style - quotation style mainly. Apart from that, this looks good to merge to me.

mbaruh committed:
Merge branch 'main' into new-filter-schema

mbaruh committed:
Make the unique constraint reversible

Co-authored-by: Amrou <amrbellalouna@gmail.com>

mbaruh committed:
Documentation improvements and fixes

Co-authored-by: Amrou <amrbellalouna@gmail.com>

mbaruh committed:
Merge the extra kwargs creation functions

Co-authored-by: Amrou <amrbellalouna@gmail.com>

mbaruh committed:
Specify the common elements in the validation errors

Co-authored-by: Amrou <amrbellalouna@gmail.com>

mbaruh committed:
Use consistent quoting style

Co-authored-by: Johannes Christ <jc@jchri.st>

jchristgit reviewed (approved):
Thank you!

mbaruh committed:
Rename additional_field to additional_settings

wookie184 reviewed (approved):
The changes since my last review look good, and the bot side works, so this seems good, thanks!

mbaruh committed:
Add maximum auto-timeout duration validation

shtlrs reviewed (approved):
One more for extra luck 🍀

mbaruh committed:
Make additional_settings non-null with dict default

This makes sure that the value in the DB is always a valid JSON, ensuring the unique constraint will work properly.
